File: public/cpp/024-remove-nth-node-from-end.cpp
/*
 * LeetCode Problem #19: Remove Nth Node From End of List
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/remove-nth-node-from-end-of-list/
 * 
 * Problem Statement:
 * Given the head of a linked list, remove the nth node from the end of the list
 * and return its head.
 * 
 * Example:
 * Input: head = [1,2,3,4,5], n = 2
 * Output: [1,2,3,5]
 * Explanation: Remove the 2nd node from the end (node with value 4).
 * 
 * Follow up: Could you do this in one pass?
 */

#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    // ==================== APPROACH 1: Two Pointer (One Pass) ====================
    /*
     * Algorithm:
     * - Use two pointers with n gap between them
     * - Move both pointers until fast reaches end
     * - Slow pointer will be at node before target
     * - Remove target node
     * 
     * Time Complexity: O(L) where L = length of list
     * Space Complexity: O(1)
     * 
     * When to use: This is the OPTIMAL solution
     * 
     * Key Insight:
     * - Maintain n+1 gap to position slow before target
     * - Dummy node handles edge case of removing head
     * - Single pass through the list
     */
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        
        ListNode* fast = dummy;
        ListNode* slow = dummy;
        
        // Move fast n+1 steps ahead
        for (int i = 0; i <= n; i++) {
            fast = fast->next;
        }
        
        // Move both pointers until fast reaches end
        while (fast != nullptr) {
            fast = fast->next;
            slow = slow->next;
        }
        
        // Remove the nth node from end
        ListNode* toDelete = slow->next;
        slow->next = slow->next->next;
        delete toDelete;
        
        ListNode* result = dummy->next;
        delete dummy;
        return result;
    }
    
    // ==================== APPROACH 2: Two Pass (Calculate Length) ====================
    /*
     * Algorithm:
     * - First pass: calculate list length
     * - Second pass: go to (length - n)th node
     * - Remove the target node
     * 
     * Time Complexity: O(L)
     * Space Complexity: O(1)
     * 
     * When to use: When two passes are acceptable
     * 
     * Key Insight:
     * - Nth from end = (length - n + 1)th from start
     * - More straightforward but requires two passes
     */
    ListNode* removeNthFromEnd_TwoPass(ListNode* head, int n) {
        // Calculate length
        int length = 0;
        ListNode* current = head;
        while (current != nullptr) {
            length++;
            current = current->next;
        }
        
        // Handle edge case: remove head
        if (length == n) {
            ListNode* newHead = head->next;
            delete head;
            return newHead;
        }
        
        // Find node before target
        current = head;
        for (int i = 0; i < length - n - 1; i++) {
            current = current->next;
        }
        
        // Remove target node
        ListNode* toDelete = current->next;
        current->next = current->next->next;
        delete toDelete;
        
        return head;
    }
    
    // ==================== APPROACH 3: Recursion ====================
    /*
     * Algorithm:
     * - Recursively traverse to end
     * - Count nodes on return path
     * - Remove node when count equals n
     * 
     * Time Complexity: O(L)
     * Space Complexity: O(L) for recursion stack
     * 
     * When to use: When recursive solution preferred
     * 
     * Key Insight:
     * - Count from end during backtracking
     * - Return modified list from recursion
     */
    ListNode* removeNthFromEnd_Recursive(ListNode* head, int n) {
        int count = 0;
        return removeHelper(head, n, count);
    }
    
private:
    ListNode* removeHelper(ListNode* node, int n, int& count) {
        if (node == nullptr) {
            return nullptr;
        }
        
        node->next = removeHelper(node->next, n, count);
        count++;
        
        // If this is the nth node from end, skip it
        if (count == n) {
            ListNode* toDelete = node;
            node = node->next;
            delete toDelete;
        }
        
        return node;
    }
    
public:
    // ==================== APPROACH 4: Stack ====================
    /*
     * Algorithm:
     * - Push all nodes onto stack
     * - Pop n nodes to reach target
     * - Remove target node
     * 
     * Time Complexity: O(L)
     * Space Complexity: O(L) for stack
     * 
     * When to use: When stack-based solution is natural
     * 
     * Key Insight:
     * - Stack reverses order (LIFO)
     * - Easy to access nth from end
     * - Uses extra space
     */
    ListNode* removeNthFromEnd_Stack(ListNode* head, int n) {
        stack<ListNode*> stk;
        ListNode* dummy = new ListNode(0);
        dummy->next = head;
        
        // Push all nodes including dummy
        ListNode* current = dummy;
        while (current != nullptr) {
            stk.push(current);
            current = current->next;
        }
        
        // Pop n nodes to reach node before target
        for (int i = 0; i < n; i++) {
            stk.pop();
        }
        
        // Remove target node
        ListNode* prev = stk.top();
        ListNode* toDelete = prev->next;
        prev->next = prev->next->next;
        delete toDelete;
        
        ListNode* result = dummy->next;
        delete dummy;
        return result;
    }
};

// ==================== HELPER FUNCTIONS ====================
ListNode* createList(vector<int>& values) {
    if (values.empty()) return nullptr;
    
    ListNode* head = new ListNode(values[0]);
    ListNode* current = head;
    
    for (int i = 1; i < values.size(); i++) {
        current->next = new ListNode(values[i]);
        current = current->next;
    }
    
    return head;
}

void printList(ListNode* head) {
    cout << "[";
    while (head != nullptr) {
        cout << head->val;
        if (head->next != nullptr) cout << ",";
        head = head->next;
    }
    cout << "]";
}

void deleteList(ListNode* head) {
    while (head != nullptr) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }
}

// ==================== TEST CASES ====================
void runTests() {
    Solution sol;
    
    // Test Case 1: Remove from middle
    // Input: head = [1,2,3,4,5], n = 2
    // Expected: [1,2,3,5]
    cout << "Test Case 1: Remove from middle\n";
    vector<int> vals1 = {1, 2, 3, 4, 5};
    ListNode* head1 = createList(vals1);
    ListNode* result1 = sol.removeNthFromEnd(head1, 2);
    cout << "Output: ";
    printList(result1);
    cout << "\nExpected: [1,2,3,5]\n\n";
    deleteList(result1);
    
    // Test Case 2: Remove head (only one node)
    // Input: head = [1], n = 1
    // Expected: []
    cout << "Test Case 2: Remove only node\n";
    vector<int> vals2 = {1};
    ListNode* head2 = createList(vals2);
    ListNode* result2 = sol.removeNthFromEnd_TwoPass(head2, 1);
    cout << "Output: ";
    printList(result2);
    cout << "\nExpected: []\n\n";
    
    // Test Case 3: Remove head (multiple nodes)
    // Input: head = [1,2], n = 2
    // Expected: [2]
    cout << "Test Case 3: Remove head\n";
    vector<int> vals3 = {1, 2};
    ListNode* head3 = createList(vals3);
    ListNode* result3 = sol.removeNthFromEnd_Recursive(head3, 2);
    cout << "Output: ";
    printList(result3);
    cout << "\nExpected: [2]\n\n";
    deleteList(result3);
    
    // Test Case 4: Remove last node
    // Input: head = [1,2,3], n = 1
    // Expected: [1,2]
    cout << "Test Case 4: Remove last node\n";
    vector<int> vals4 = {1, 2, 3};
    ListNode* head4 = createList(vals4);
    ListNode* result4 = sol.removeNthFromEnd_Stack(head4, 1);
    cout << "Output: ";
    printList(result4);
    cout << "\nExpected: [1,2]\n\n";
    deleteList(result4);
    
    // Test Case 5: Two nodes, remove first
    // Input: head = [1,2], n = 1
    // Expected: [1]
    cout << "Test Case 5: Two nodes, remove last\n";
    vector<int> vals5 = {1, 2};
    ListNode* head5 = createList(vals5);
    ListNode* result5 = sol.removeNthFromEnd(head5, 1);
    cout << "Output: ";
    printList(result5);
    cout << "\nExpected: [1]\n\n";
    deleteList(result5);
}

int main() {
    runTests();
    return 0;
}

/*
 * COMPARISON OF APPROACHES:
 * 
 * 1. Two Pointer One Pass (RECOMMENDED):
 *    Time: O(L), Space: O(1)
 *    Pros: Optimal, single pass, constant space
 *    Cons: Requires careful pointer management
 *    Best for: Most interview scenarios
 * 
 * 2. Two Pass with Length:
 *    Time: O(L), Space: O(1)
 *    Pros: Straightforward, easy to understand
 *    Cons: Requires two passes
 *    Best for: When clarity is priority
 * 
 * 3. Recursion:
 *    Time: O(L), Space: O(L)
 *    Pros: Elegant, counts from end naturally
 *    Cons: Extra space for recursion stack
 *    Best for: When recursion preferred
 * 
 * 4. Stack:
 *    Time: O(L), Space: O(L)
 *    Pros: Easy to access from end
 *    Cons: Extra space for stack
 *    Best for: Stack-based thinking
 * 
 * INTERVIEW TIPS:
 * - Clarify if n is always valid (1 ≤ n ≤ length)
 * - Mention the one-pass requirement
 * - Explain the two-pointer technique with gap
 * - Discuss why dummy node is useful
 * - Walk through example showing pointer movement
 * - Handle edge case: removing head
 * - Consider edge case: single node list
 * - Mention memory cleanup (delete nodes)
 * - Discuss trade-offs between approaches
 * - Consider follow-up: what if n could be invalid?
 * 
 * KEY INSIGHTS:
 * - Two pointers with n+1 gap positions slow before target
 * - Dummy node simplifies head removal case
 * - Nth from end = (length - n + 1)th from start
 * - One pass is possible with two pointers
 * - Fast pointer reaches null when slow is at target
 * 
 * STEP-BY-STEP WALKTHROUGH:
 * List: 1->2->3->4->5, n = 2 (remove 4)
 * 
 * Initial: dummy->1->2->3->4->5
 * fast = dummy, slow = dummy
 * 
 * Move fast n+1=3 steps:
 *   fast at 3, slow at dummy
 * 
 * Move both until fast is null:
 *   fast at 4, slow at 1
 *   fast at 5, slow at 2
 *   fast at null, slow at 3
 * 
 * slow->next is target (4)
 * Remove: slow->next = slow->next->next
 * Result: 1->2->3->5
 * 
 * COMMON MISTAKES:
 * - Not using dummy node (complicates head removal)
 * - Moving fast only n steps (should be n+1)
 * - Not handling single node case
 * - Not checking if n is valid
 * - Memory leaks from not deleting nodes
 * - Off-by-one errors in pointer positioning
 * - Not considering removing head
 * - Incorrect gap between pointers
 * - Forgetting to delete dummy node
 * - Not handling null input
 * 
 * FOLLOW-UP QUESTIONS:
 * - What if n could be larger than list length?
 * - How would you remove multiple nodes?
 * - Can you do it without dummy node?
 * - How would you remove nth from start?
 * - What if list is doubly linked?
 * 
 * RELATED PROBLEMS:
 * - LeetCode #237: Delete Node in a Linked List
 * - LeetCode #203: Remove Linked List Elements
 * - LeetCode #83: Remove Duplicates from Sorted List
 * - LeetCode #82: Remove Duplicates from Sorted List II
 * - LeetCode #2095: Delete the Middle Node of a Linked List
 */

// Made with Bob

File: src/data/tutorials/024-remove-nth-node-from-end.js
/**
 * Tutorial content for #024 — Remove Nth Node From End
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given the head of a linked list and an integer n, remove the nth node from the end of the list and return the head. Try to do it in one pass.`,
    example: `list: 1→2→3→4→5, n=2\n→ 2nd from end is node(4)\n→ remove it: 1→2→3→5\n✅ Answer: 1→2→3→5`,
    keyInsight: `Use two pointers with a gap of n nodes between them. When the fast pointer reaches the end, the slow pointer is exactly at the node before the one to remove.`,
  },

  approaches: {
    'Two Pointer One Pass': {
      intuition: `Send a "fast" pointer n+1 steps ahead of a "slow" pointer. Move both at the same speed until fast reaches the end of the list. Now slow is positioned just before the node to remove. Use a dummy head to handle the edge case where the first node itself is removed.`,
      steps: [
        `Create dummy = new ListNode(0, head). Set fast = dummy, slow = dummy.`,
        `Advance fast n+1 times (so there's a gap of n between fast and slow).`,
        `Move both fast and slow forward together until fast is null.`,
        `slow.next = slow.next.next (skip the nth-from-end node).`,
        `Return dummy.next.`,
      ],
      example: `list: 1→2→3→4→5, n=2\ndummy→1→2→3→4→5\n\nAdvance fast 3 steps (n+1=3):\n  fast: dummy→1→2→3\n  slow: at dummy\n\nMove both until fast=null:\n  fast=3→4, slow=1\n  fast=4→5, slow=2\n  fast=5→null... wait, fast should reach null:\n  fast=3→4→5→null: move 3 times from 3:\n  Actually from position 3: fast→4, slow→1; fast→5, slow→2; fast→null, slow→3\n  slow=3, slow.next=4 (the target)\n\nslow.next = slow.next.next = 5\n✅ Result: 1→2→3→5`,
      keyInsight: `O(n) time (one pass), O(1) space. The n+1 gap ensures slow stops exactly one node before the target. The dummy head elegantly handles removal of the very first node (n == list length).`,
    },

    'Two Pass Length': {
      intuition: `First pass: count the total length of the list. Second pass: calculate the position from the front (length - n), then walk to that node and remove the next one. Simple and easy to reason about, but requires two traversals.`,
      steps: [
        `First pass: count the length L of the list.`,
        `Create dummy = new ListNode(0, head). Set curr = dummy.`,
        `Compute the number of steps to the node before the target: steps = L - n.`,
        `Advance curr exactly steps times.`,
        `curr.next = curr.next.next (remove the target node).`,
        `Return dummy.next.`,
      ],
      example: `list: 1→2→3→4→5, n=2\n\nPass 1: length L=5\nSteps to node before target: 5-2=3\n\nPass 2 (from dummy):\n  dummy→1 (step 1) → 2 (step 2) → 3 (step 3) → curr=3\n  curr.next = 4, curr.next.next = 5\n  curr.next = 5 → list becomes 1→2→3→5\n✅ Answer: 1→2→3→5`,
      keyInsight: `O(n) time (two passes), O(1) space. More intuitive than the two-pointer approach — easier to explain. The two-pointer one-pass solution is O(n) the same, just more clever.`,
    },

    'Recursion': {
      intuition: `Recurse all the way to the end of the list, then count nodes on the way back. When the counter reaches n on the return path, that node is the one to remove — return its next instead of itself. Use a dummy head so the head node itself can be removed uniformly.`,
      steps: [
        `Create dummy = new ListNode(0, head). Set count = [0] (mutable integer wrapper).`,
        `Define helper(node): if node == null return null; otherwise node.next = helper(node.next).`,
        `Inside helper, increment count after the recursive call returns.`,
        `If count[0] == n, skip this node by returning node.next instead of node.`,
        `Call dummy.next = helper(head) and return dummy.next.`,
      ],
      example: `list: 1→2→3→4→5, n=2\n\nRecurse to end:\nhelper(null) → return null, count stays 0\nhelper(5) → count=1, 1≠2, return node(5)\nhelper(4) → count=2, 2==2 → return node(4).next = node(5) [node 4 removed]\nhelper(3) → count=3, return node(3)→node(5)\nhelper(2) → count=4, return node(2)→node(3)→node(5)\nhelper(1) → count=5, return node(1)→node(2)→node(3)→node(5)\n✅ Result: 1→2→3→5`,
      keyInsight: `O(n) time, O(n) space for the call stack. Elegant but uses stack depth proportional to list length. The count is passed as a single-element array to allow mutation inside the recursive helper.`,
    },

    'Stack': {
      intuition: `Push every node (including the dummy head) onto a stack. Popping n nodes takes you right past the node to remove. The next peek reveals the predecessor node — set its next pointer to skip the target.`,
      steps: [
        `Create dummy = new ListNode(0, head). Push dummy and every subsequent node onto a stack.`,
        `Pop exactly n nodes from the top of the stack (discarding them).`,
        `Peek at the top — this is the predecessor of the node to remove.`,
        `Set predecessor.next = predecessor.next.next to unlink the target.`,
        `Return dummy.next.`,
      ],
      example: `list: 1→2→3→4→5, n=2\nStack (top→bottom): [5, 4, 3, 2, 1, dummy]\n\nPop 1st: node(5)\nPop 2nd: node(4) — this is the node to remove\nPeek: node(3) — this is the predecessor\nnode(3).next = node(3).next.next = node(5)\n✅ Result: 1→2→3→5`,
      keyInsight: `O(n) time, O(n) space for the stack. Conceptually simple — the LIFO order of the stack naturally exposes nodes from the end first. Popping n nodes to find the nth-from-end is a direct mapping of the problem statement.`,
    },
  },
}
